flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -0,0 +1,462 @@
+			currentResourceRequirements.add(ResourceRequirement.create(resourceRequirement.getKey(), resourceRequirement.getValue()));
+		}
+
+		return currentResourceRequirements;
+	}
+
+	@Override
+	public Collection<SlotOffer> offerSlots(
+		Collection<? extends SlotOffer> offers,
+		TaskManagerLocation taskManagerLocation,
+		TaskManagerGateway taskManagerGateway,
+		long currentTime) {
+
+		LOG.debug("Received {} slot offers from TaskExecutor {}.", offers.size(), taskManagerLocation);
+		final Collection<SlotOffer> acceptedSlotOffers = new ArrayList<>();
+		final Collection<SlotOffer> candidates = new ArrayList<>();
+
+		// filter out already accepted offers
+		for (SlotOffer offer : offers) {
+			final AllocationID allocationId = offer.getAllocationId();
+			if (slotPool.containsSlot(allocationId)) {
+				acceptedSlotOffers.add(offer);
+			} else {
+				candidates.add(offer);
+			}
+		}
+
+		final Collection<SlotOfferMatching> matchings = matchOffersWithOutstandingRequirements(candidates);
+
+		final Collection<AllocatedSlot> acceptedSlots = new ArrayList<>();
+		ResourceCounter acceptedResources = ResourceCounter.empty();
+
+		for (SlotOfferMatching matching : matchings) {
+			if (matching.getMatching().isPresent()) {
+				final ResourceProfile matchedResourceProfile = matching.getMatching().get();
+
+				final AllocatedSlot allocatedSlot = createAllocatedSlot(
+					matching.getSlotOffer(),
+					taskManagerLocation,
+					taskManagerGateway);
+
+				acceptedSlots.add(allocatedSlot);
+				acceptedSlotOffers.add(matching.getSlotOffer());
+
+				acceptedResources = acceptedResources.add(matchedResourceProfile, 1);
+
+				// store the ResourceProfile against which the given slot has matched for future book-keeping
+				slotToRequirementProfileMappings.put(allocatedSlot.getAllocationId(), matchedResourceProfile);
+			}
+		}
+
+		slotPool.addSlots(acceptedSlots, currentTime);
+		increaseAvailableResources(acceptedResources);
+
+		if (!acceptedSlots.isEmpty()) {
+			notifyNewSlots.accept(acceptedSlots);
+		}
+
+		return acceptedSlotOffers;
+	}
+
+	private Collection<SlotOfferMatching> matchOffersWithOutstandingRequirements(Collection<SlotOffer> slotOffers) {
+		ResourceCounter unfulfilledResources = calculateUnfulfilledResources();
+
+		final Collection<SlotOfferMatching> matching = new ArrayList<>();
+
+		for (SlotOffer slotOffer : slotOffers) {
+			ResourceProfile matchingResourceProfile = null;
+
+			if (unfulfilledResources.containsResource(slotOffer.getResourceProfile())) {
+				unfulfilledResources = unfulfilledResources.subtract(slotOffer.getResourceProfile(), 1);
+
+				// use the profile of the slot offer since it is a direct match against the requirement
+				matchingResourceProfile = slotOffer.getResourceProfile();
+			} else {
+				for (ResourceProfile unfulfilledResource : unfulfilledResources.getResources()) {
+					if (slotOffer.getResourceProfile().isMatching(unfulfilledResource)) {
+						matchingResourceProfile = unfulfilledResource;
```

Comment:
hmm...I think the idea was to accept any slot that could be used for _some_ requirement if they don't exactly match, to prevent cases where we reject a slot because some other slot has taken it's place. For example, assume the requirements are one small and one large slot. Large slot offer comes first, fulfills small requirement, small slot arrives and is rejected.

Reply:
I think the idea was not to accept excess slots which we are sure not to use. Which slots, if there are multiple matching ones, to accept is a different question. I think that can be a follow up question and I'd be fine to simply take the first matching one.